src/python_testing/TC_DEM_2_2.py

```diff
@@ -0,0 +1,325 @@
+            TestStep(
+                "14", "TH sends PowerAdjustRequest with Power=PowerAdjustmentCapability[0].MaxPower, Duration=PowerAdjustmentCapability[0].MinDuration, Cause=LocalOptimization. Verify Command response is ConstraintError"),
+            TestStep("15", "TH sends TestEventTrigger command to General Diagnostics Cluster on Endpoint 0 with EnableKey field set to PIXIT.DEM.TEST_EVENT_TRIGGER_KEY and EventTrigger field set to PIXIT.DEM.TEST_EVENT_TRIGGER for User Opt-out Grid Optimization Test Event. Event DEM.S.E01(PowerAdjustEnd) sent with Cause=UserOptOut, Duration= approx time from step 11 to step 15, EnergyUse= a valid value"),
+            TestStep("15a", "TH reads ESAState attribute. Verify value is 0x01 (Online)"),
+            TestStep("15b", "TH reads OptOutState attribute. Verify value is 0x03 (OptOut)"),
+            TestStep("15c", "TH reads PowerAdjustmentCapability attribute. Value has to include Cause=NoAdjustment."),
+            TestStep("16", "TH sends TestEventTrigger command to General Diagnostics Cluster on Endpoint 0 with EnableKey field set to PIXIT.DEM.TEST_EVENT_TRIGGER_KEY and EventTrigger field set to PIXIT.DEM.TEST_EVENT_TRIGGER for User Opt-out Test Event Clear"),
+            TestStep("16a", "TH reads ESAState attribute. Verify value is 0x01 (Online)"),
+            TestStep("16b", "TH reads OptOutState attribute. Verify value is 0x00 (NoOptOut)"),
+            TestStep(
+                "17", "TH sends PowerAdjustRequest with Power=PowerAdjustmentCapability[0].MaxPower, Duration=PowerAdjustmentCapability[0].MinDuration, Cause=LocalOptimization. Verify Command response is Success and event DEM.S.E00(PowerAdjustStart) sent"),
+            TestStep("17a", "TH reads ESAState attribute. Verify value is 0x04 (PowerAdjustActive)"),
+            TestStep("17b", "TH reads PowerAdjustmentCapability attribute. Value has to include Cause=LocalOptimizationAdjustment."),
+            TestStep("18", "Wait 10 seconds. Event DEM.S.E01(PowerAdjustEnd) sent with Cause=NormalCompletion, Duration=10s, EnergyUse= a valid value"),
+            TestStep("18a", "TH reads ESAState attribute. Verify value is 0x01 (Online)"),
+            TestStep("18b", "TH reads PowerAdjustmentCapability attribute. Value has to include Cause=NoAdjustment."),
+            TestStep("19", "TH sends TestEventTrigger command to General Diagnostics Cluster on Endpoint 0 with EnableKey field set to PIXIT.DEM.TEST_EVENT_TRIGGER_KEY and EventTrigger field set to PIXIT.DEM.TEST_EVENT_TRIGGER for Power Adjustment Test Event Clear"),
+        ]
+
+        return steps
+
+    @async_test_body
+    async def test_TC_DEM_2_2(self):
+        # pylint: disable=too-many-locals, too-many-statements
+        """Run the test steps."""
+        # These values have to correlate with the values configured in
+        # DeviceEnergyManagementManufacturerImpl::SetTestEventTrigger_PowerAdjustment()
+        min_power = 5 * 1000 * 1000
```

Comment:
Do we need to know these in the test script? The test plan says that the values can be anything, but the min_duration needs to be 10s.

Reply:
It just felt a tighter test checking against an explicit value. Will change as requested above.